### PR TITLE
Expr: remove redundant fields, add nrExprs

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

Minor things I've found while helping out a bit with #8778

- Slim int and float exprs
- Count how many AST nodes we load 

# Context

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by Ericson2314** [comment]
Is there a reason why we don't do the same thing for the other exprs, like `ExprString`?

**Comment by roberth** [comment]
> ExprString

That one seemed to be in an optimum, although possibly a local one of course. It uses a bunch of `std::move` and it appears that the bytes are already shared.
Removing the `std::string` from the Expr would remove the thing that owns the bytes. Another solution is perhaps to make it leak those bytes and than refer to it only in the `Value`. This does not seem like an obvious win, and I'm only doing low hanging fruit in this PR.

The bools are `ExprVar`, so not much that can be done there.

`ExprPath` is like string but is subject to change anyway with `SourcePath` I think.

Didn't spot any other instances of such duplication.

**Comment by Ericson2314** [comment]
Ah! I didn't realize the `std::string` owns what the `Value` points too. OK that makes perfect sense and makes clear rules for when the other field is needed.

File: src/libexpr/nixexpr.hh [hunk 1]
```diff
@@ -155,6 +155,10 @@ std::string showAttrPath(const SymbolTable & symbols, const AttrPath & attrPath)
 
 struct Expr
 {
+    static unsigned long nrExprs;
+    inline Expr() {
```
Comment (inclyc) [inline src/libexpr/nixexpr.hh:159]:
```suggestion
    Expr() {
```

this is redundant since definitions in C++ record is implicitly `inline`-ed

**Review by edolstra — APPROVED** [review]

**Review by inclyc — APPROVED** [review]

**Pull request merged**